### ISSUE TITLE
conversion of big int to string

### ISSUE THREAD

**Issue by afabri** [boostorg/multiprecision#617]
We use   `i.convert_to<std::string>`  for a multi precision integer `i` and observe a factor of 10 difference for the `cpp_int` backend compared to the `gmp_int` backend.    Is there a better way than using `convert_to()` ?

**Comment by jzmaddock** [comment]
Indeed, plus is there any particular reason not to just call the `.str()` method?

**Comment by afabri** [comment]
I have put a self contained program [here](https://gist.github.com/afabri/db497cd574b959c110f0a6adbfefb9f7)  on gist.github.com.    Calling `.str()`  has the same difference in performance.

**Comment by afabri** [comment]
Under vtune I can see that it is the `divide_unsigned_helper()` which is time consuming.

**Comment by jzmaddock** [comment]
Ah performance, not value of result, I misunderstood, apologies.

Yes we know that cpp_int division is terribly slow compared to GMP, but GMP is a tour-de-force of optimisation and we're not really trying to compete on that front (which is not to say we couldn't do better).

**Comment by afabri** [comment]
In our use case (a multiprecision float with mantissa and exponent being arbitrary precision ints)  we are only interested in the leading N characters of the string representing the number.

**Comment by ckormanyos** [comment]
> In our use case (a multiprecision float with mantissa and exponent being arbitrary precision ints) we are only interested in the leading N characters of the string representing the number.

Cound you do something a bit non-elegant like get the MSB (for $2^n$}), then get the highest order limb and maks a talbe-based addition for the leading bits? It does not really sound like a lot of fun, but might work?

**Comment by afabri** [comment]
> Ah performance, not value of result, I misunderstood, apologies.
> 
> Yes we know that cpp_int division is terribly slow compared to GMP, but GMP is a tour-de-force of optimisation and we're not really trying to compete on that front (which is not to say we couldn't do better).

@jzmaddock Can you give me a pointer to a publication that tells what has to be implemented to catch up?

**Comment by ckormanyos** [comment]
> to a publication that tells what has to be implemented to catch up?

I'll put out a few details. `cpp_int` division is known to be slow. What could be done?

  - Investigate Knuth long-division. I have an implementation of it (but with limbs ordered the other way around). Knuth long-division is quite difficult to program but it is rather quick for low limb counts.
  - One of the things that can be done, but we are probably a long way away from, would be to visualize integr division as floating-point multiplication via numerator scaling-up and subsequent Newton-Raphson iteration and re-rationalization. If you convert `cpp_int` to `cpp_bin_float` this would be a very roundabout way to start seeing if this could help. Why? Besauce at least we go up to Karatsuba in multiplication.

@jzmaddock might add more.

P.S. I like the algorithm descriptions in:

[1] R. Brent and P. Zimmermann, "Modern Computer Arithmetic", Cambridge University Press, 2011.
[2] J. Arndt, "Matters Computational", Springer Verlag, 2011.
[3] and of course the Knuth's volume 2.
